packages/edit-navigation/src/store/actions.js

```diff
@@ -0,0 +1,274 @@
+/**
+ * External dependencies
+ */
+import { invert, keyBy, omit } from 'lodash';
+import { v4 as uuid } from 'uuid';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import {
+	getNavigationPost,
+	getPendingActions,
+	isProcessingPost,
+	getMenuItemToClientIdMapping,
+	resolveMenuItems,
+	dispatch,
+	apiFetch,
+} from './controls';
+import { menuItemsQuery } from './utils';
+
+/**
+ * Creates a menu item for every block that doesn't have an associated menuItem.
+ * Requests POST /wp/v2/menu-items once for every menu item created.
+ *
+ * @param {Object} post A navigation post to process
+ * @return {Function} An action creator
+ */
+export const createMissingMenuItems = serializeProcessing( function* ( post ) {
+	const menuId = post.meta.menuId;
+
+	const mapping = yield {
+		type: 'GET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+		postId: post.id,
+	};
+	const clientIdToMenuId = invert( mapping );
+
+	const stack = [ post.blocks[ 0 ] ];
+	while ( stack.length ) {
+		const block = stack.pop();
+		if ( ! ( block.clientId in clientIdToMenuId ) ) {
+			const menuItem = yield apiFetch( {
+				path: `/__experimental/menu-items`,
+				method: 'POST',
+				data: {
+					title: 'Placeholder',
+					url: 'Placeholder',
+					menu_order: 0,
+				},
+			} );
+
+			mapping[ menuItem.id ] = block.clientId;
+			const menuItems = yield resolveMenuItems( menuId );
+			yield dispatch(
+				'core',
+				'receiveEntityRecords',
+				'root',
+				'menuItem',
+				[ ...menuItems, menuItem ],
+				menuItemsQuery( menuId ),
+				false
+			);
+		}
+		stack.push( ...block.innerBlocks );
+	}
+
+	yield {
+		type: 'SET_MENU_ITEM_TO_CLIENT_ID_MAPPING',
+		postId: post.id,
+		mapping,
+	};
+} );
+
+/**
+ * Converts all the blocks into menu items and submits a batch request to save everything at once.
+ *
+ * @param {Object} post A navigation post to process
+ * @return {Function} An action creator
+ */
+export const saveNavigationPost = serializeProcessing( function* ( post ) {
+	const menuId = post.meta.menuId;
+	const menuItemsByClientId = mapMenuItemsByClientId(
+		yield resolveMenuItems( menuId ),
+		yield getMenuItemToClientIdMapping( post.id )
+	);
+
+	try {
+		const response = yield* batchSave(
+			menuId,
+			menuItemsByClientId,
+			post.blocks[ 0 ]
+		);
+		if ( ! response.success ) {
+			throw new Error();
```

Comment:
Is throwing necessary? Looks like we can use `if ( response.success )` to determine what action we dispatch.

Reply:
There are two different types of error that may happen:

1. Response has code 200 and body indicating `success === false`. In this case `apiFetch` just returns the response for us to handle.
2. Something unpredicted happens, like HTTP response code 500. In this case `apiFetch` throws an error.

For now we handle errors by just saying "there was an error" without communicating the root cause. Throwing allows us to handle both types of errors in the `catch` clause. I think it's fine, at least until we implement more sophisticated/granular error handling.

Reply:
> I think it's fine, at least until we implement more sophisticated/granular error handling.

Can you open an issue for this?